### PR TITLE
Add pretrained AlexNet

### PR DESCRIPTION
**Pull request by SurajK7**
This PR implements pretrained Alexnet and all other necessary functions needed to implement it.

### CODE REVIEW

**Comment by championballer** [comment]
Did you get a chance to run inference using this on some popular dataset? Would love to know how the performance was.

**Comment by SurajK7** [comment]
@championballer No, not yet, but since I have transferred weights directly from PyTorch model which was trained on ImageNet so the performance should be [similiar](https://pytorch.org/docs/stable/torchvision/models.html#classification), by performance you mean performance on metrics like top5-accuracy, right? Or something along the line of inference time and such because I am not sure about the latter.

**Comment by championballer** [comment]
Yeah, I meant in terms of inference time.  

On a side note, do you think it would be better to somehow run this on a sample picture of the provided dimensions, to check its label? That would be a better check of the correct loading and performance of the model.

**Comment by austinvhuang** [comment]
LGTM

File: examples/alexNet/Main.hs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        <*> sample (conv2)
+        <*> sample (conv3)
+        <*> sample (conv4)
+        <*> sample (conv5)
+        <*> sample (fc1)
+        <*> sample (fc2)
+        <*> sample (fc3)
+
+an :: AN -> Tensor -> Tensor
+an AN{..} input =
```
Comment (apaszke) [inline examples/alexNet/Main.hs:63]: nit: Can we please use `alexnet` and `AlexNet` instead of `an`?

File: examples/alexNet/Main.hs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+instance Randomizable ANSpec AN where
+    sample ANSpec {..} = AN 
+        <$> sample (conv1)
+        <*> sample (conv2)
+        <*> sample (conv3)
+        <*> sample (conv4)
+        <*> sample (conv5)
+        <*> sample (fc1)
+        <*> sample (fc2)
+        <*> sample (fc3)
```
Comment (apaszke) [inline examples/alexNet/Main.hs:60]: Note to self: We should write generics for `Randomizable` for modules that don't have parameters

File: hasktorch/src/Torch/Functional.hs [hunk 3]
```diff
@@ -551,9 +551,14 @@ maxPool2d
     -> Bool -- ^ ceil mode
     -> Tensor -- ^ input
     -> Tensor -- ^ output
-maxPool2d kernelSize stride padding dilation ceilMode self =
+maxPool2d (kernelSize0, kernelSize1) (stride0, stride1) (padding0, padding1) (dilation0, dilation1) ceilMode self =
```
Comment (apaszke) [inline hasktorch/src/Torch/Functional.hs:554]: Wouldn't it be nicer to have a helper like `asList :: (Int, Int) -> [Int]` and simply apply that to every argument without explicit unpacking?

File: hasktorch/src/Torch/NN.hs [hunk 4]
```diff
@@ -139,3 +139,34 @@ instance Parameterized Linear
...
+conv2d'' layer stride padding input = 
+  Torch.Functional.conv2d' w b stride padding input
+    where
+        w = toDependent (conv2dWeight layer)
+        b = toDependent (conv2dBias layer)
+
+instance Randomizable Conv2dSpec Conv2d where
+  sample Conv2dSpec{..} = do
+      w <- makeIndependent =<< randnIO' [outputChannelSize, inputChannelSize, kernelSize0, kernelSize1]
+      b <- makeIndependent =<< randnIO' [outputChannelSize]
```
Comment (apaszke) [inline hasktorch/src/Torch/NN.hs:167]: Can you please copy the initialization method from PyTorch? This one is unlikely to yield good results.

File: hasktorch/src/Torch/NN.hs [hunk 5]
```diff
@@ -139,3 +139,34 @@ instance Parameterized Linear
...
+      return $ Conv2d w b
+
+instance Parameterized Conv2d
+
+instance Parameterized [Conv2d]
```
Comment (apaszke) [inline hasktorch/src/Torch/NN.hs:172]: I don't think we should be generating this instance FWIW

**Comment by apaszke** [comment]
Also, this is great! Thanks for implementing it!

File: examples/alexNet/Main.hs (hunk 1 above)
Comment (SurajK7) [reply examples/alexNet/Main.hs:63]: Yeah, sure.

File: hasktorch/src/Torch/Functional.hs (hunk 3 above)
Comment (SurajK7) [reply hasktorch/src/Torch/Functional.hs:554]: Yeah, sure will get this done.

File: hasktorch/src/Torch/NN.hs (hunk 4 above)
Comment (SurajK7) [reply hasktorch/src/Torch/NN.hs:167]: Actually, I copied it from the typed implementation of Conv2d, maybe I should change it there as well?

File: hasktorch/src/Torch/NN.hs (hunk 5 above)
Comment (SurajK7) [reply hasktorch/src/Torch/NN.hs:172]: Sure will remove it.

Comment (SurajK7) [reply hasktorch/src/Torch/NN.hs:172]: I think, for loading and updating parameters of a layer, it is necessary for the layer to be an instance of `Parameterized` class.  Or maybe I am missing something? Feel free to correct me.

Comment (SurajK7) [reply hasktorch/src/Torch/NN.hs:172]: I think, for loading and updating parameters of an untyped layer, it is necessary for the layer to be an instance of Parameterized class, given my implementation of alexnet or I'll have to modify the implementation a bit. Or maybe I am missing something? Feel free to correct me.

File: examples/alexNet/Main.hs [hunk 6]
```diff
@@ -83,9 +88,23 @@ an AN{..} input =
...
     $ input
 
+normalize :: Either String Tensor -> Tensor
+normalize (Right img) = cat 0 [r', g', b']
+    where
+        img' = divScalar (255.0::Float) $ toType D.Float (hwc2chw img)
+        [r, g, b] = I.split (reshape [3, 224, 224] img') 1 0 
+        r' = divScalar (0.229::Float) (subScalar(0.485::Float) r) 
+        g' = divScalar (0.224::Float) (subScalar(0.456::Float) g) 
+        b' = divScalar (0.225::Float) (subScalar(0.406::Float) b)
```
Comment (SurajK7) [inline examples/alexNet/Main.hs:98]: I am not sure if this the best way to normalize. Any suggestions? @junjihashimoto

File: hasktorch/src/Torch/NN.hs [hunk 7]
```diff
@@ -163,8 +163,9 @@ conv2d'' layer stride padding input =
 
 instance Randomizable Conv2dSpec Conv2d where
   sample Conv2dSpec{..} = do
-      w <- makeIndependent =<< randnIO' [outputChannelSize, inputChannelSize, kernelSize0, kernelSize1]
-      b <- makeIndependent =<< randnIO' [outputChannelSize]
+      w <- makeIndependent =<< kaimingUniform FanIn (LeakyRelu $ Prelude.sqrt 5.0) [outputChannelSize, inputChannelSize, kernelSize0, kernelSize1]
+      uniform <- randIO' [outputChannelSize]
+      b <- makeIndependent =<< pure (mulScalar ((2::Int) * kernelSize0 * kernelSize1) (subScalar (0.5 :: Float) uniform))
```
Comment (SurajK7) [inline hasktorch/src/Torch/NN.hs:168]: @apaszke Is this initialization fine?

**Comment by SurajK7** [comment]
Update: As of now we can run inference on images of dimensions 224 * 224 pixels and of format PNG, bitmap and JPEG.

File: examples/alexNet/Labels.hs [hunk 8]
```diff
@@ -0,0 +1,1007 @@
+module Labels (labels) where
+
+import Data.Array
+
+labels = array (0, 999) $ zip [0 .. 999] list
+
+list =
```
Comment (apaszke) [inline examples/alexNet/Labels.hs:7]: I'm confused. Why do we have this huge file now?

File: hasktorch/src/Torch/NN.hs (hunk 5 above)
Comment (apaszke) [reply hasktorch/src/Torch/NN.hs:173]: I don't see why would you need this instance for `AlexNet`, because you never use a list of `Conv2D`s? Also, if you really need it, then declare the instance in your code and not in the core library.

File: hasktorch/src/Torch/NN.hs [hunk 9]
```diff
@@ -139,3 +139,35 @@ instance Parameterized Linear
 --     return $ Linear{..}
 
 instance Parameterized [Linear]
+
+data Conv2dSpec = 
+  Conv2dSpec {
+    inputChannelSize  :: Int, 
+    outputChannelSize :: Int, 
+    kernelSize0       :: Int, 
+    kernelSize1       :: Int
```
Comment (apaszke) [inline hasktorch/src/Torch/NN.hs:148]: Maybe instead of `0` and `1` suffixes call them `kernelHeight` and `kernelWidth`?

File: hasktorch/src/Torch/NN.hs [hunk 10]
```diff
@@ -139,3 +139,35 @@ instance Parameterized Linear
...
+    } deriving (Show, Eq)
+
+data Conv2d = 
+  Conv2d { 
+    conv2dWeight :: Parameter, 
+    conv2dBias   :: Parameter
+    } deriving (Show, Generic)
+
+conv2d'' :: Conv2d -> (Int, Int) -> (Int, Int) -> Tensor -> Tensor
+conv2d'' layer stride padding input =
```
Comment (apaszke) [inline hasktorch/src/Torch/NN.hs:158]: Why does this have two apostrophes?

File: hasktorch/src/Torch/NN.hs (hunk 7 above)
Comment (apaszke) [reply hasktorch/src/Torch/NN.hs:168]: Weight looks good, but bias seems to be missing a square root?

File: hasktorch/src/Torch/NN.hs (hunk 5 above)
Comment (SurajK7) [reply hasktorch/src/Torch/NN.hs:173]: Sorry, I misunderstood I thought you were talking about the `Conv2d` instance. Yeah, it doesn't make sense to have the list instance there.

File: examples/alexNet/Labels.hs (hunk 8 above)
Comment (SurajK7) [reply examples/alexNet/Labels.hs:7]: Actually, the file contains the labels of 1000 classes from ImageNet dataset. For now, I have placed it in the alexNet directory maybe later we can move it to Torch.Vision, or if unnecessary remove it.

File: hasktorch/src/Torch/NN.hs (hunk 10 above)
Comment (SurajK7) [reply hasktorch/src/Torch/NN.hs:158]: The `Torch` module exports both `Torch.NN` and `Torch.Functional` modules and there are already definitions for `conv2d` and `conv2d'` present in `Torch.Functional` module hence the two apostrophes.

File: examples/alexNet/Labels.hs (hunk 8 above)
Comment (austinvhuang) [reply examples/alexNet/Labels.hs:7]: Would prefer to keep data files out of the repo and out of the history as well so may need to rebase. We can have scripts to download data and metadata as needed, but let's try not to have dataset contents themselves in the repo unless there's a good reason.

File: hasktorch/src/Torch/NN.hs (hunk 10 above)
Comment (austinvhuang) [reply hasktorch/src/Torch/NN.hs:158]: We haven't formalized it quite yet, but a convention we've started using on functions which implement the 'forward' inference on data with paramaterized instances is "xxxForward". For example, have been doing that in this PR with RNNs https://github.com/hasktorch/hasktorch/pull/319/files

Comment (SurajK7) [reply hasktorch/src/Torch/NN.hs]: I have made the necessary changes. Could you please have a look?

File: examples/alexNet/Labels.hs (hunk 8 above)
Comment (apaszke) [reply examples/alexNet/Labels.hs]: Also, it's usually good practice to avoid adding things "just in case", because they might not ever end up being useful and you can accumulate a bunch of dead code this way. Just something to keep in mind for the future!

Comment (SurajK7) [reply examples/alexNet/Labels.hs]: Thank you for your suggestion, I will abide by it.

**Comment by SurajK7** [comment]
I am thinking of adding a fine-tuning example based on AlexNet. Any suggestions for the data-set to work upon, I am thinking of [Oxford-IIIT Pet Dataset](http://www.robots.ox.ac.uk/~vgg/data/pets/). Thoughts?

**Comment by junjihashimoto** [comment]
Let's try fine-tuning.
I think we have the problem of resource management on GPU.
If you have any problems, please let me know.
I hope you can solve it.

**Comment by junjihashimoto** [comment]
BTW, I think we can merge this pr and open other issue using pet-data.
@SurajK7 Should we merge this?

**Comment by championballer** [comment]
Hey guys, I have a suggestion in regards to flatten function in Functional.hs. @SurajK7 Can you use Dim datatype instead of Int? Will maintain the uniformity in the source file?

**Comment by SurajK7** [comment]
@junjihashimoto Yeah, we can merge this PR, I will open a separate draft PR for the pet data-set example.
Also, @championballer thank you for pointing out I have done the needful.

File: hasktorch/src/Torch/Functional.hs [hunk 11]
```diff
@@ -566,7 +566,15 @@ maxPool2d
...
-        self kernelSize stride padding dilation ceilMode
+        self 
+        (asList kernelSize)
+        (asList stride)
+        (asList padding)
+        (asList dilation)
+        ceilMode
+        where
+            asList :: (Int, Int) -> [Int]
+            asList (a0, a1) = [a0, a1]
```
Comment (austinvhuang) [inline hasktorch/src/Torch/Functional.hs:577]: I've run into this elsewhere, would be preferable if we could add tuples to the casting mechanism although in this context it's a pretty minor thing (any thoughts @junjihashimoto ?)

Anyway out of scope to have to resolve in this PR.
